File: clients/typescript/src/stream-rpc.test.js
import assert from "node:assert/strict";
import test from "node:test";

import {
  OresRpcStreamClient,
  RpcStreamTransportError,
} from "./stream-rpc.js";

function asyncFrames(frames) {
  return {
    async *[Symbol.asyncIterator]() {
      for (const frame of frames) yield frame;
    },
  };
}

test("prepare performs zero I/O; stream is the sole open boundary", async () => {
  let opens = 0;
  const framedStream = {
    carrier: "websocket",
    async open(call) {
      opens += 1;
      assert.equal(call.key, "demo.events.watch_stream");
      assert.equal(call.id, "s-1");
      return {
        incoming: asyncFrames([
          { id: "s-1", t: "data", body: { value: 1 } },
          { id: "s-1", t: "data", body: { value: 2 } },
          { id: "s-1", t: "end" },
        ]),
      };
    },
  };
  const client = new OresRpcStreamClient({
    framedStream,
    operations: ["demo.events.watch_stream"],
    idPrefix: "s-",
  });

  const call = client
    .prepare(
      "demo.events.watch_stream",
      { method: "GET", path: "/v1/events" },
      (value) => value.value,
    )
    .addQueryField("room_id", 7)
    .withBody({ include_deleted: false });

  assert.equal(opens, 0);
  const stream = await call.stream();
  assert.equal(opens, 1);

  const values = [];
  for await (const value of stream) values.push(value);
  assert.deepEqual(values, [1, 2]);
  assert.equal(stream.context.ended, true);
  assert.equal(stream.context.cancelled, false);
});

test("a stream builder can open only once", async () => {
  const framedStream = {
    carrier: "tcp",
    async open(call) {
      return { incoming: asyncFrames([{ id: call.id, t: "end" }]) };
    },
  };
  const call = new OresRpcStreamClient({
    framedStream,
    operations: ["demo.watch_stream"],
  }).prepare("demo.watch_stream", { method: "GET", path: "/v1/watch" });

  await call.stream();
  await assert.rejects(
    call.stream(),
    (error) =>
      error instanceof RpcStreamTransportError &&
      error.reason === "protocol",
  );
});

test("correlation mismatches fail closed", async () => {
  const framedStream = {
    carrier: "websocket",
    async open() {
      return {
        incoming: asyncFrames([{ id: "wrong", t: "data", body: 1 }]),
      };
    },
  };
  const stream = await new OresRpcStreamClient({
    framedStream,
    operations: ["demo.watch_stream"],
    idPrefix: "expected-",
  })
    .prepare("demo.watch_stream", { method: "GET", path: "/v1/watch" })
    .stream();

  await assert.rejects(
    async () => {
      for await (const _ of stream) {
        void _;
      }
    },
    (error) =>
      error instanceof RpcStreamTransportError &&
      error.reason === "protocol",
  );
});

test("remote errors retain code and reason", async () => {
  const framedStream = {
    carrier: "tcp",
    async open(call) {
      return {
        incoming: asyncFrames([
          { id: call.id, t: "error", code: "denied", message: "nope" },
        ]),
      };
    },
  };
  const stream = await new OresRpcStreamClient({
    framedStream,
    operations: ["demo.watch_stream"],
  })
    .prepare("demo.watch_stream", { method: "GET", path: "/v1/watch" })
    .stream();

  await assert.rejects(
    async () => {
      for await (const _ of stream) {
        void _;
      }
    },
    (error) =>
      error instanceof RpcStreamTransportError &&
      error.reason === "remote" &&
      error.code === "denied",
  );
});

test("cancel delegates once", async () => {
  let cancels = 0;
  const framedStream = {
    carrier: "websocket",
    async open() {
      return {
        incoming: asyncFrames([]),
        async cancel() {
          cancels += 1;
        },
      };
    },
  };
  const stream = await new OresRpcStreamClient({
    framedStream,
    operations: ["demo.watch_stream"],
  })
    .prepare("demo.watch_stream", { method: "GET", path: "/v1/watch" })
    .stream();

  await stream.cancel();
  await stream.cancel();
  assert.equal(cancels, 1);
  assert.equal(stream.context.cancelled, true);
});

File: clients/dart/test/stream_rpc_test.dart
import 'package:ores_api_docs/ores_api_docs.dart';
import 'package:test/test.dart';

final class _Session implements RpcStreamSession {
  _Session(this.frames);

  final List<RpcStreamFrame> frames;
  int cancels = 0;

  @override
  Stream<RpcStreamFrame> get incoming =>
      Stream<RpcStreamFrame>.fromIterable(frames);

  @override
  Future<void> cancel() async {
    cancels += 1;
  }
}

final class _Wire implements FramedRpcStream {
  _Wire(this.carrier, this.session);

  @override
  final RpcStreamCarrier carrier;
  final _Session session;
  late _Session openedSession;
  int opens = 0;

  @override
  Future<RpcStreamSession> open(RpcStreamCall call) async {
    opens += 1;
    final rebound = session.frames.map((frame) {
      switch (frame.kind) {
        case RpcStreamFrameKind.data:
          return RpcStreamFrame.data(id: call.id, body: frame.body);
        case RpcStreamFrameKind.end:
          return RpcStreamFrame.end(id: call.id);
        case RpcStreamFrameKind.error:
          return RpcStreamFrame.error(
            id: call.id,
            code: frame.code ?? 'unknown',
            message: frame.message,
          );
        case RpcStreamFrameKind.cancel:
          return RpcStreamFrame.cancel(id: call.id);
        case RpcStreamFrameKind.call:
          return RpcStreamFrame.call(id: call.id);
      }
    }).toList();
    openedSession = _Session(rebound);
    return openedSession;
  }
}

void main() {
  test('prepare/configure is inert until stream()', () async {
    final wire = _Wire(
      RpcStreamCarrier.websocket,
      _Session([
        RpcStreamFrame.data(id: '', body: {'value': 1}),
        RpcStreamFrame.data(id: '', body: {'value': 2}),
        RpcStreamFrame.end(id: ''),
      ]),
    );
    final client = OresRpcStreamClient(
      stream: wire,
      operations: const ['demo.events.watch_stream'],
      idPrefix: 's-',
    );
    final builder = client
        .prepare<int>(
          'demo.events.watch_stream',
          const RpcStreamRequest(method: 'GET', path: '/v1/events'),
          (value) => (value! as Map)['value'] as int,
        )
        .addQueryField('room_id', 7)
        .withBody({'active': true});

    expect(wire.opens, 0);
    final stream = await builder.stream();
    expect(wire.opens, 1);
    expect(await stream.values.toList(), [1, 2]);
    expect(stream.context.ended, isTrue);
  });

  test('stream builder can open only once', () async {
    final wire = _Wire(
      RpcStreamCarrier.tcp,
      _Session([RpcStreamFrame.end(id: '')]),
    );
    final builder = OresRpcStreamClient(
      stream: wire,
      operations: const ['demo.watch_stream'],
    ).prepare<int>(
      'demo.watch_stream',
      const RpcStreamRequest(method: 'GET', path: '/v1/watch'),
      (value) => value as int,
    );

    await builder.stream();
    await expectLater(builder.stream(), throwsStateError);
  });

  test('cancellation delegates once', () async {
    final wire = _Wire(RpcStreamCarrier.tcp, _Session(const []));
    final stream = await OresRpcStreamClient(
      stream: wire,
      operations: const ['demo.watch_stream'],
    ).prepare<int>(
      'demo.watch_stream',
      const RpcStreamRequest(method: 'GET', path: '/v1/watch'),
      (value) => value as int,
    ).stream();

    await stream.cancel();
    await stream.cancel();
    expect(wire.openedSession.cancels, 1);
    expect(stream.context.cancelled, isTrue);
  });
}
